library/core/src/future/poll_fn.rs

```diff
@@ -24,6 +24,75 @@ use crate::task::{Context, Poll};
 /// assert_eq!(read_future.await, "Hello, World!".to_owned());
 /// # }
 /// ```
+///
+/// ## Capturing a pinned state
+///
+/// Example of a closure wrapping an inner future:
+///
+/// ```
+/// # async fn run() {
+/// use core::future::{Future, poll_fn};
+/// use core::mem;
+/// use core::task::Poll;
+///
+/// /// Wraps a `Future` so that it yields `Pending` one time before forwarding to the inner
+/// /// `Future`.
+/// fn yield_once_then<T>(fut: impl Future<Output = T>) -> impl Future<Output = T> {
+///     let mut should_yield = true;
+///     let mut pinned_fut = Box::pin(fut);
+///     poll_fn(move |cx| if mem::take(&mut should_yield) {
+///         // Remember: for a `Future` impl to be correct, any `Pending` return has to have,
+///         // once ready, a matching `wake` call, lest it not be rescheduled for execution.
```

Comment:
This is certainly necessary for this function to work as yield_once_then, but IIUC the explanation about yield_once_then is unrelated to the purpose of this section and distracting.

Is it possible to make the example simpler? Personally, I think an example that simply polls the future in a closure would be sufficient.

Reply:
(Drive-by nitpicking; don't take this too seriously.)

Arguably, this example is already too simple — _because_ there is an enclosing `async` block, and `should_yield` never becomes true after it is false, it is possible to rewrite

```rust
let mut pinned_fut = pin!(fut);
poll_fn(move |cx| if mem::take(&mut should_yield) {
    cx.waker().wake_by_ref();
    Poll::Pending
} else {
    pinned_fut.as_mut().poll(cx)
}).await
```

as

```rust
poll_fn(move |cx| if mem::take(&mut should_yield) {
    cx.waker().wake_by_ref();
    Poll::Pending
} else {
    Poll::Ready(())
}).await;
fut.await    // no need to pin!()
```

and this is (in my opinion) better-factored code, showing how to take maximum advantage of `async {}` rather than writing any more manual-polling code than you have to. I could worry that people learning `async` internals will read this code and wonder why _not_ to write the latter.

My first idea for an example that actually needs the `pin!` and poll would be one which yields after a time interval has elapsed (adding yield points for better cooperative multitasking), but that's got perhaps a bit too much complexity.

Reply:
Yeah, that's a fair point; what about a naïve `select`, then?

That could be simple whilst legitimately polling the inner futures:

```diff
  /// Resolves to the first future that completes. In the event of a tie, `a` wins.
  fn select_biased<T>(
      a: impl Future<Output = T>,
      b: impl Future<Output = T>,
  ) -> impl Future<Output = T>
  {
+     async {
          let (mut a, mut b) =
-             (Box::pin(a), Box::pin(b))
+             (pin!(a), pin!(b))
          ;
          future::poll_fn(move |cx| {
              match (a.as_mut().poll(cx), b.as_mut().poll(cx)) {
                  (Poll::Ready(r), _) | (_, Poll::Ready(r)) => Poll::ready(r),
                  _ => Poll::Pending,
              }
          })
+         .await
+     }
  }
```

Reply:
I like that example much better.

One further nitpick, though: it doesn't just let `a` win, it also _polls `b` and discards the result if any_, which good `select`s don't do and which will introduce cancellation bugs (e.g. if you are `select`ing among two message channels, this will drop some messages from the second). It's easy to fix at the cost of 3 more lines (but might be more readable anyway):

```rust
if let Poll::Ready(r) = a.as_mut().poll(cx) {
    Poll::Ready(r)
} else if let Poll::Ready(r) = b.as_mut().poll(cx) {
    Poll::Ready(r)
} else {
    Poll::Pending
}
```

(Thanks for taking my input. I'm bothering to write all this because in my experience, beginners _will_ read examples and assume that everything in them is good practice even when it's not about the thing the example is for.)

Reply:
Started implementing this as of https://github.com/rust-lang/rust/pull/109970/commits/9541287e315d15cb851e58f598b39fc3cf597723 (not on my machine, but a gh Codespace, so it will probably still need a `fmt` pass and whatnot).

FWIW, this thus puts and end to the `wake()` noise/documentation question, since it's no longer relevant.

  - (I feel like a companion PR just implementing `yield_once` using `poll_fn` could be a nice thing to add nonetheless (but without `pin!` nor the `_then` part), since I do still believe that the `wake` specificites ought to have examples. But I won't push for it myself, nor mention it anymore)